[IV-21-22]  Objetivo 6

gomares opened the pull request:
* [x] ¿Se ha evitado Circle CI, porque a estas alturas ya lo ha usado todo el
mundo?
* [x] ¿Se han establecido criterios *a priori* para elegir el sistema de
integración continua más conveniente y se han calificado los sistemas que
cumplan esos requisitos según esos criterios?
* [x] ¿Se han examinado varios sistemas de integración continua?
* [x] ¿Se han configurado varios sistemas de integración continua?
* [x] ¿Se ha configurado correctamente el *Checks API* en los sistemas en los que
   sea necesario para que aparezca correctamente en GitHub (y se pueda comprobar
   desde los tests)?
* [x] ¿Uno de los sistemas configurados permite comprobar cuales son las versiones
   dellenguaje con las que funciona correctamente nuestra aplicación?
* [x] ¿Se escogen de forma adecuada las versiones del lenguaje que se testean,
   tanto en el sistema de CI como en el contenedor Docker?
* [x] ¿Se han justificado correctamente las versiones del lenguaje que se están
testeando y se ha comprobado que no se comprueban varias veces lo mismo (la
misma versión en CI y en el contenedor Docker en otro sistema CI?


En este Pull Request se han trabajado en los siguientes contenidos con el objetivo de superar el objetivo 6:

- Elección de dos sistemas de integración continua.

- Justificación de los sistemas de CI elegidos.

- Configuración de los sistemas de CI y justificación para pasar los tests automáticamente al hacer push al repositorio y comprobar las versiones del lenguaje con las que funciona el proyecto.

Se ha trabajado en las issues:

- Sistemas de Integración continua #34 .

- Configuración de los sistemas de integración continua #35 .

- Añadir Sistemas de integración continua a iv.yaml #36 .

gomares committed:
Closes #34. Docs con imagenes

gomares committed:
Closes #36 . Fichero IV clave añadida

gomares committed:
Avanza #35 . Configurando GH Actions

gomares committed:
Ajustando versiones Python

gomares committed:
Ajusto version checkout

gomares committed:
Closes #35. Configura CircleCI

gomares committed:
Updated config.yml

gomares committed:
Añado espacio

gomares committed:
Merge branch 'Objetivo-6' of https://github.com/gomares/More-mangas into Objetivo-6

gomares committed:
Reformulo y cambio de nombre a docker_run_tests

gomares committed:
Update update_docker.yml

gomares committed:
Update update_docker.yml

JJ commented on .github/workflows/update_docker.yml:
```diff
@@ -1,14 +1,22 @@
+      - 'pyproject.toml'
+      - 'Dockerfile'
+    branches:
+      - 'main'
```
¿Sólo se actualiza en main?

gomares replied:
Voy a realizar cambios para que no solo se actualice en main si no en las ramas pertencientes a objetivos de la asignatura.

JJ replied:
¿En main no?

JJ commented on .github/workflows/continous_integration.yml:
```diff
@@ -0,0 +1,18 @@
+name: Docker Image CI
```
Podías llamarlo de otra forma...

gomares replied:
Voy a pensar en un título más apropiado para lo que hago.

JJ replied:
... y el nombre del fichero...

JJ commented on .github/workflows/continous_integration.yml:
```diff
@@ -0,0 +1,18 @@
+
+on:
+  push:
+    branches: [ main ]
```
Por favor, mirad bien qué significa lo de las branches. No lo va a pasar en un PR, por ejemplo.

gomares replied:
Cierto.

JJ reviewed (changes requested):
Aparte de lo dicho en privado, tienes que darle una vuelta a esto.

gomares committed:
Nuevo título para continous_integrations.yml

gomares committed:
Añado branches de hitos para pasar los tests ahí también

gomares commented:
Cambios realizados.

JJ commented:
Revisores → @jesusmarzor @amerigal @modejota

amerigal commented on .circleci/config.yml:
```diff
@@ -0,0 +1,27 @@
+    jobs:
+      - tests:
+          matrix: # Matriz de versiones
+            parameters: 
+              version: ["3.8","3.9","3.10"]
```
Creo que sería interesante una justificación de por qué son esas las versiones que se están testeando.

modejota replied:
Same here, aunque viendo que el 99% de los proyectos lo utilizan, me hago una idea

gomares replied:
Ya he dejado una justificación, muchas gracias chicos. En el objetivo 5 también hablé acerca de por qué incluí la nueva versión de Python. 😄

amerigal reviewed (commented):
Muy completo y trabajado el objetivo :1st_place_medal: 

Solo hacer unas observaciones. Por un lado, con respecto a los workflows de github actions, si vas a incluir todas las ramas diría que no es necesario que incluyas la condición de branches. Además, me da la impresión de que tal y como está, los workflows se están ejecutando dos veces, una vez por ser un push a una rama del tipo Objetivo-* y otra por tratarse de un pull request a main. No estoy del todo seguro, porque aparecen muchos tests ejecutados y lo mismo es por otra cosa, pero por si puedes revisar eso para no hacer ejecuciones innecesasarias.

modejota commented:
Por mi parte, 3/4 lo mismo que comenta @amerigal.
- Si especificas que se lance en los push y pull request puede que se ejecute dos veces innecesariamente, que tampoco sé si realmente pasa así. En mis pruebas los puse en una u otra cosa, pero no en ambas.
- Si especificas que se lance en "main" y "Objetivo-*", básicamente lo estamos lanzando siempre, ya que no utilizamos otros nombres de rama, por lo que casi que no es necesario especificarlo.
- Justificar un poco porque se utiliza cada versión. Puedo intuir que 3.8 por que será la mínima que soporte alguna funcionalidad concreta, 3.10 por ser la más reciente, y 3.9 porque sí.

JJ commented:
⛹ Revisores → @amerigal @paula1999 @Olasergiolas

Olasergiolas commented on docs/objetivo-6.md:
```diff
@@ -0,0 +1,24 @@
+
+* En primer lugar tenemos a **TravisCI**, recomendado por el profesor por ser muy intuitivo y con una interfaz buena para implementar la integración continua a los proyectos fácil y rápido. Intenté probar a usar la versión trial pero además de solo durar un mes me pedía tarjeta de crédito y no se por qué pero al ponerla me aparecía un mensaje de la app del banco diciendo que habían denegado un pa ...
+
+* En segundo lugar descarté **SempahoreCI** por la misma razón que Travis, exigen que pagues pasados [14 días](https://semaphoreci.com/pricing). Además de esto no es tan popular como otras opciones de pago y que me gustan más como el propio Travis.
```
No usaría "popular" para la justificación sino las consecuencias que tiene el que sea más popular, es decir, podrías poner en su lugar que tiene una mayor comunidad o que problemas comunes con la plataforma están mejor documentados.

gomares replied:
Sí, tienes razón. SemaphoreCI la descarté por tener versión de pago y una documentación peor que la de Travis, este me pareció la opción a usar si estaba dispuesto a pagar.

Olasergiolas commented on docs/objetivo-6.md:
```diff
@@ -0,0 +1,24 @@
+
+* En segundo lugar descarté **SempahoreCI** por la misma razón que Travis, exigen que pagues pasados [14 días](https://semaphoreci.com/pricing). Además de esto no es tan popular como otras opciones de pago y que me gustan más como el propio Travis.
+
+* Las dos últimas opciones que investigué eran por un lado **Jenkins** y por otro **CircleCI**. Jenkins presentaba unos [problemas](https://circleci.com/migrate-jenkins-to-circleci/) que CircleCI no, es cierto que es más completo por tener otras funcionalidades que conseguimos añadiendo plugins (tiene más de 1200 plugins). No elegí Jenkins debido a que a la hora de mantener el producto es mucho m ...
```
Typo en "cause".

gomares replied:
Corregida

Olasergiolas commented on docs/objetivo-6.md:
```diff
@@ -0,0 +1,24 @@
+
+* En segundo lugar descarté **SempahoreCI** por la misma razón que Travis, exigen que pagues pasados [14 días](https://semaphoreci.com/pricing). Además de esto no es tan popular como otras opciones de pago y que me gustan más como el propio Travis.
+
+* Las dos últimas opciones que investigué eran por un lado **Jenkins** y por otro **CircleCI**. Jenkins presentaba unos [problemas](https://circleci.com/migrate-jenkins-to-circleci/) que CircleCI no, es cierto que es más completo por tener otras funcionalidades que conseguimos añadiendo plugins (tiene más de 1200 plugins). No elegí Jenkins debido a que a la hora de mantener el producto es mucho m ...
```
Enlace markdown "editar tu archivo config.yml" roto.

gomares replied:
Arreglado

Olasergiolas commented on .github/workflows/continous_integration.yml:
```diff
@@ -0,0 +1,22 @@
+name: Docker from DockeHub run tests 
```
Typo en "DockeHub".

gomares replied:
Corregida

sgIOlas reviewed (changes requested):
En general lo veo bien, aunque creo que deberías concretar algo más lo que has hecho para el objetivo en el PMV, además de que por ejemplo dices que solo se incorpora un sistema de integración. Échale un vistazo también a un par de fallos tontos que te he comentado 👍.

gomares committed:
Cambio en las versiones de la matriz.

Elimino la 3.10 ya que la testeo con el workflow de GH Action, añado la 3.7

gomares committed:
Renombro automatizacion en GH Action y cambio los triggers

Ahora se realiza el test cuando hay un push y se elimina la opción de pull request ya que se estaba ejecutando dos veces, tenia razón amerigal.

gomares committed:
Corrección de typos y justificación de versiones de la matriz.

Corrige algunos errores tipográficos indicados por revisores y también otros vistos por mí mismo y añado un anexo para justificar las matrices que testear.

gomares committed:
Corrijo fichero de actualizar docker

No tenía sentido tener un push y un pull_request cuando lo podía simplificar dejando solo un push y especificando solo los archivos referentes al dockerfile y dependencias, sin necesidad de especificar las ramas.

gomares committed:
Corrección de un typo

gomares commented:
Estuve investigando las branches y los pushes y pull requests como eventos de workflows y ya he cambiado los ficheros de los workflows para que ahora solo se ejecute una vez los tests y no dos como antes hacía. @amerigal @modejota . Muchas gracias por la aportación.

He mejorado algunas cosillas que me indicaste @Olasergiolas, gracias a ti también 😄 .
También he corregido lo que me dijo el profesor en su día cuando tenía problemas para añadirme revisores jejeje.

Creo que está todo en orden, si me dais algo más de feedback lo pongo como listo para revisión.

amerigal commented:
> Estuve investigando las branches y los pushes y pull requests como eventos de workflows y ya he cambiado los ficheros de los workflows para que ahora solo se ejecute una vez los tests y no dos como antes hacía. @amerigal @modejota . Muchas gracias por la aportación.
> 
> He mejorado algunas cosillas que me indicaste @Olasergiolas, gracias a ti también smile . También he corregido lo que me dijo el profesor en su día cuando tenía problemas para añadirme revisores jejeje.
> 
> Creo que está todo en orden, si me dais algo más de feedback lo pongo como listo para revisión.

Veo bien los cambios realizados :+1:. Lo único que no me termina de convencer de esa parte es la justificación de que se usara la versión 3.7 inicialmente porque es la que estaba instalada en el ordenador. Pienso que las justificaciones deben ir más bien enfocadas a las necesidades del proyecto.

Lo último comentar sobre los issues y el milestone. Entiendo que sería apropiado que los issues introducidos se refirieran directa o indirectamente a HUs, y que esas HUs a cuya consecución están contribuyendo los issues se incluyeran también en el milestone.

gomares committed:
Closes #42. Justificación de versiones mejorada.

gomares commented:
La historia creada no era necesaria, tenía #34 ya así que voy a reescribir algunas cosas para que guarden relación entre sí. A ver si logro lo que dices @amerigal .

gomares commented:
Ya está corregido @amerigal.

sgIOlas reviewed (approved):
Por mi parte creo también que ya es suficiente para pasar el objetivo.

amerigal reviewed (approved):
Todo en orden por mi parte :+1:

gomares commented:
Perfecto, pues gracias por vuestras revisiones. Por mi parte lo veo listo para revisión @JJ. 😃

JJ commented on .github/workflows/run_tests_in_test_container.yml:
```diff
@@ -0,0 +1,14 @@
+    steps:
+      - uses: actions/checkout@v2
+      - name: Run tests
+        run: docker run -t -v `pwd`:/app/test gomares/more-mangas
```
¿Qué versión testa este?

gomares replied:
La 3.10, lo especifique cuando se creó el contenedor en objetivos anteriores, aclararé esto en la neuva documentación.

JJ commented on docs/objetivo-6.md:
```diff
@@ -0,0 +1,30 @@
+## Elección de servicio de Integración Continua
+
+Se necesitan dos servicios de integración continua para cumplir el hito. Se han valorado ciertos CIs:
```
Lo primero que hay que poner son los criterios de elección. Si se han "valorado", tienes que decir en qué se basa. ¿Están recomendados como las mejores prácticas en algún artículo?

gomares replied:
Pondré criterios preliminares en la nueva documentación. Gracias por la ayuda.

JJ commented on docs/objetivo-6.md:
```diff
@@ -0,0 +1,30 @@
+
+* En primer lugar tenemos a **TravisCI**, recomendado por el profesor por ser muy intuitivo y con una interfaz buena para implementar la integración continua a los proyectos fácil y rápido. Intenté probar a usar la versión trial pero además de solo durar un mes me pedía tarjeta de crédito y no se por qué pero al ponerla me aparecía un mensaje de la app del banco diciendo que habían denegado un pa ...
+
+* En segundo lugar descarté **SempahoreCI** por la misma razón que Travis, exigen que pagues pasados [14 días](https://semaphoreci.com/pricing). Además de esto no tiene una comunidad tan grande como otras opciones de pago y que me gustan más como el propio Travis. Es muy parecido a CircleCI , hasta el punto de no encontrar ninguna diferencia importante más allá de las diferencias en los ficheros  ...
```
Razón que tenías que haber establecido de antemano.

JJ replied:
... aparte de escribirlo correctamente.

JJ commented on docs/objetivo-6.md:
```diff
@@ -0,0 +1,30 @@
+
+## Estructura del objetivo:
+
+Elegí CircleCI y GH Actions para la realización del objetivo. Por tanto, se han configurado dos tareas relativas a la Integración Continua:
```
Como el 99% de la clase. Y por las mismas razones: son lo que habías decidido desde el principio.

JJ reviewed (changes requested):
En este objetivo se busca que sepáis buscar y elegir el sistema de CI más adecuado a vuestro proyecto, estableciendo criterios sólidos para los mismos. También que sepáis testear diferentes versiones y por qué lo hacéis No has justificado las versiones de Python empleadas, y no has justificado por qué eliges esos dos sistemas, sólo a posteriori. Me temo que no superas el objetivo.

gomares committed:
Creación de fichero de AppVeyor

Primer intento de ejecutar contenedor con versión 3.10 de Python desde AppVeyor.

gomares committed:
Comentamos el script con explicaciones

gomares committed:
Creación de travis.yml

gomares committed:
Incluyo versión 3.7 de Python ya que tiene mantenimiento

gomares committed:
Añado fichero semaphoreCI

Que se note que trabajé y probé este sistemaaaa

gomares committed:
Actualizo iv.yaml con Travis en vez de Circleci

gomares committed:
Actualizar imágenes del repositorio. Necesarias para la documentación.

gomares committed:
Closes #34. Nueva documentación.

Nueva estructura del objetivo.

*Requisitos
*Análisis de los candidatos
*Justificación sobre las versiones testeadas.
*Estructura final del objetivo
*Problemas que surgieron al ir probando distintos CIs.

JJ commented:
⛹ Revisores → @morevi @Pablont98 @lentes4k

gomares commented:
Listo para revisión @JJ. No creo que nadie quiera revisarlo 😢

JJ commented on .appveyor.yml:
```diff
@@ -0,0 +1,13 @@
+image: Ubuntu
```
Tercer fichero que me encuentro para exactamente el mismo sistema de CI y con prácticamene las mismas órdenes. CircleCI no es malo de por si, es malo porque al haberlo usado ya n personas, la persona n+1 no tenía más que copiarlo. Se trata de que aprendáis y entendáis lo que estás haciendo, no de que pongáis n sitios de integración continua.

JJ reviewed (approved):
:+1: porque has probado un montón de sistemas de CI. Pero la versión sigue siendo incorrecta, y podrías haber establecido algún criterio restrictivo de antemano.
